### DISCUSSION TITLE
Did the linkerd-multicluster-link chart purposely stop being published?

### DISCUSSION THREAD

**Discussion by wondersd** [linkerd/linkerd2#10554] [Q&A] [answered]
cant find any more recent versions of the https://github.com/linkerd/linkerd2/tree/main/multicluster/charts/linkerd-multicluster-link helm chart other than 2.9.7 in the stable helm repo (doesnt seem to have any versions published in edge repo at all)

It seems back in 2.9.x the chart was in a different location in git repo and has since was moved around, maybe publishing of it got lost in the shuffle? It seems to still be receiving updates in github, just not being published.

**Comment by alpeb** [comment, marked as answer]
The `linkerd-multicluster-link` chart is not operational on its own and that's why we stopped publishing it. It's just  an internal dependency of the `linkerd-multicluster` chart, that it embeds when it's published.

**Reply by wondersd** [reply]
@alpeb Thanks for the quick response!

Hmm, just pulled apart the linkerd-multicluster chart contents and dont see the linkerd-multicluster-link chart templates or a referenced subchart? I'm looking at `https://helm.linkerd.io/stable/linkerd-multicluster-2.11.2.tgz` (also took a look at 2.11.5, the last stable release in `https://helm.linkerd.io/stable/index.yaml`) for my current version.

Im also noticing that there isnt a linkerd-multicluster published for any 2.12.x version.

My specific use case istrying to use helm charts to perform cluster linking:

 * linkerd @ 2.11.2
 * linkerd-multicluster @ 2.11.2
 * linkerd-multicluster-link @ 2.11.2 (installed from source control)
 * explicitly defined `Link` CR and cluster secret
 
All 4 seem to be required and they do seem to work. If there is a way to get whats in `linkerd-multicluster-link` from the `linkerd-multicluster` chart Im missing how.

**Reply by alpeb** [reply]
My bad, `linkerd-multicluster-link` is _not_ embedded into `linkerd-multicluster`. `linkerd-multicluster-link` is just used as a template by the `linkerd multicluster link` command, which also adds the Link and the cluster Secret which would be hard to generate via just a helm chart. So we instruct people to rely on the CLI command only when linking clusters. But I see you've worked around that by using `linkerd-multicluster-link` and then generating the Link and Secret yourself. We can make the chart public again, with the instructions on how to generate the missing resources (open for folks to contribute on that :wink: ).

You can see all the releases for the `linkerd-multicluster` chart here:
https://artifacthub.io/packages/helm/linkerd2/linkerd-multicluster
Note that with 2.12.0 we started leveraging Helm's two versions system:
- "Application Version": which you'll see the latest is `stable-2.12.4`
- "Chart Version": which the latest is `30.2.6`
That way we can detach the evolution of the charts from the app itself.

**Reply by wondersd** [reply]
Ah I missed the departure of versions between chart an app.

If that chart can be published again that'd be great! I can update on generating the CR and secret

**Reply by whiskeysierra** [reply]
I'll sketch an alternative approach that I successfully applied to link clusters without manually invoking the CLI.
What I wanted was a full IaC setup using terraform, incl. all the resources that make up a link.

Let me just name the important ingredients here, so that others that stumble over this can pick and choose what makes sense in their setup:

- Terraform external datasource (https://registry.terraform.io/providers/hashicorp/external/latest/docs/data-sources/external)
- A custom script (see below) that invokes the `linkerd multicluster cli` command and returns the YAML output
- Dockerized Linkerd CLI to have full control over the linkerd version
- Terraform kubectl_file_documents datasource (https://registry.terraform.io/providers/gavinbunney/kubectl/latest/docs/data-sources/kubectl_file_documents) to translate the multi-document YAML string into a terraform data structure
- Terraform kubectl_manifest resource (https://registry.terraform.io/providers/gavinbunney/kubectl/latest/docs/resources/kubectl_manifest) to actually create and manage each resource

What this gives us:
- ✅ Independent of some internal implementation details of the linkerd cli (e.g. the linkerd-multicluster-link helm chart)
- ✅ fully managed link resources using IaC
- ✅ no human operator with privileged access required

```bash
#!/bin/bash

set -euo pipefail

local_cluster=
remote_cluster=
kubeconfig=
linkerd_version=
eval "$(jq -r '@sh "local_cluster=\(.local_cluster_name) remote_cluster=\(.remote_cluster_name) kubeconfig=\(.kube_config) linkerd_version=\(.linkerd_version)"')"

# create .kubeconfig for target cluster
echo "${kubeconfig}" > ".kubeconfig-${remote_cluster}"

# create link
docker run \
  -w "${PWD}" \
  -v "${PWD}:${PWD}" \
  "cr.l5d.io/linkerd/cli-bin:${linkerd_version}" \
  multicluster link \
  --kubeconfig ".kubeconfig-${remote_cluster}" \
  --cluster-name "${remote_cluster}" \
  --service-account-name "linkerd-service-mirror-remote-access-${local_cluster}" \
  --set "enableHeadlessServices=true" \
   | jq --slurp --raw-input '{content: .}'
```
